---
title: "Analysis of aDMRs differentially expressed irrespectively of treatment"
author: "Mikhail G. Dozmorov"
date: "August 17, 2015"
output: html_document
---
```{r setup, echo=FALSE, message=FALSE, warning=FALSE}
suppressMessages(source("../../../GenomeRunner/R.GenomeRunner/utils1.R"))
# Set up the environment
library(knitr) 
opts_chunk$set(cache.path='cache/', fig.path='img/', cache=F, tidy=T, fig.keep='high', echo=F, dpi=100, warnings=F, message=F, comment=NA, warning=F) #out.width=700, 
options(replace.assign=TRUE)#, width=120)
library(pander)
panderOptions('table.split.table', Inf)
set.seed(1)
library(xlsx)
library("ReactomePA")
library("DOSE")
library(KEGG.db)
```

08/07/2015 - the significant DMR results for all treatments and for all treatments combined. 
```{r}
load("/Users/mikhail/Documents/Work/VCU_work/Coleen/Breast_cancer/data/Tim/DMR_CpG.Rdata")
# write.table(data.frame(paste("chr", temp.all$chr, sep=""), temp.all$start, temp.all$stop), "/Users/mikhail/Documents/Work/VCU_work/Coleen/Breast_cancer/data/Tim/DMR_CpG.bed", sep="\t", row.names=FALSE, col.names=FALSE, quote=F)
```

Examine length distribution.

```{r}
hist(temp.all$length, n=100)
summary(temp.all$length)
```

Examine number of CpGs

```{r}
hist(temp.all$num_cpg, n=100)
summary(temp.all$num_cpg)
```

DMR background position 08/25/2015
---

```{r}
load("data/Tim/Mikhail_dmr_bkg.Rdata")
nrow(dmr.bkg[ dmr.bkg$length > 0, ]) # How many DMRs
sum(dmr.bkg$num_cpg) # How many CpGs total
write.table(dmr.bkg[ dmr.bkg$length > 0, c("chr", "start", "stop")], "data/Tim/DMR_CpG_all.bed", sep="\t", quote=F, row.names=F, col.names = F)
```

Start annotations
---

```{r}
library(ChIPpeakAnno)
# DMRs-global
group1.bed <- read.table("data/Tim/DMR_CpG.bed", sep="\t", stringsAsFactors = FALSE)
group1.bed <- BED2RangedData(group1.bed) # Convert to RangedData
# All DMRs 
group2.bed <- read.table("data/Tim/DMR_CpG_all.bed", sep="\t", stringsAsFactors = FALSE)
group2.bed <- BED2RangedData(group2.bed) # Convert to RangedData
# Potential annotation duplicates, not used
# dupeRows2 <- unlist(lapply(ranges(TSS.human.NCBI36), duplicated))
```

We pull transcript annotations from BiomaRt, and annotate our peaks in relation to their transcription start sites (TSSs). We also peek inside how the annotation looks like, and check what types of annotation data do we have.

```{r}
# library(biomaRt)
# mart <- useMart("ENSEMBL_MART_ENSEMBL", dataset="hsapiens_gene_ensembl", host="feb2014.archive.ensembl.org",path="/biomart/martservice",archive=FALSE, verbose=TRUE) # Last mart containing HG19 genome annotation
# myFullAnnotation = getAnnotation(mart, featureType=c("TSS","miRNA", "Exon", "5utr", "3utr", "ExonPlusUtr", "transcript")) # An example in relation to what we can get annotations
# myFullAnnotation = getAnnotation(mart, featureType=c("TSS")) # Getting annotations is time consuming
# myFullAnnotation = readRDS("data/myFullAnnotation_TSS.Rds")
data(TSS.human.NCBI36)
annotatedPeak = annotatePeakInBatch(group1.bed, AnnotationData=TSS.human.NCBI36, select = "first")
pander(head(as.data.frame(annotatedPeak))) # An overview of how annotation results look like
colnames(as.data.frame(annotatedPeak)) # Check column names in the annotation results
pie(table(as.data.frame(annotatedPeak)$insideFeature))
```

So, we have "shortestDistance" column, and can subset the annotated peaks by it. For convenience, we use designated function for it, finding peaks within 5K around TSSs.

```{r}
# Which genes are within 5K region aroung TSSs
closest <- as.data.frame(annotatedPeak)$shortestDistance < 5000
table(closest)
annotatedPeak.5K <- annotatedPeak[ closest, ]
library(biomaRt)
mart <- useMart("ensembl", dataset="hsapiens_gene_ensembl")
genes <- getBM(attributes = c('ensembl_gene_id', 'hgnc_symbol', 'description'), filters = 'ensembl_gene_id', values = annotatedPeak.5K$feature, mart = mart, uniqueRows = TRUE)
write.xlsx(left_join(data.frame(ensemblId=annotatedPeak.5K$feature, shortestDistance=annotatedPeak.5K$shortestDistance), as.data.frame(genes), by=c("ensemblId" = "ensembl_gene_id")), "results/DMR_CpG-global_genes5K.xlsx", sheetName = "genes")

library(org.Hs.eg.db)
over <- getEnrichedGO(annotatedPeak, orgAnn="org.Hs.eg.db", maxP=0.01, multiAdj=FALSE, minGOterm=10, multiAdjMethod="")
head(unique(over$bp[, c(-10)]))
write.xlsx(unique(over$bp[, c(-10)]), "/results/enrichment-DMR_CpG-global.xlsx", sheetName = "BP-peaks", row.names=F, append = T)
write.xlsx(unique(over$mf[, c(-10)]), "/results/enrichment-DMR_CpG-global.xlsx", sheetName = "MF-peaks", row.names=F, append = T)
write.xlsx(unique(over$cc[, c(-10)]), "/results/enrichment-DMR_CpG-global.xlsx", sheetName = "CC-peaks", row.names=F, append = T)
enrichedPATH <- getEnrichedPATH(annotatedPeak, orgAnn= 'org.Hs.eg.db', pathAnn='KEGG.db', feature_id_type="ensembl_gene_id", maxP=0.01, minPATHterm=10, multiAdjMethod="none")
write.xlsx(unique(enrichedPATH[, c(-2)]), "/results/enrichment-DMR_CpG-global.xlsx", sheetName = "KEGG-peaks", row.names=F, append = T)
```

Let's analyze corresponding Ensembl IDs. Not used, it's not as good as native analysis

```{r eval=FALSE}
source("/Users/mikhail/Documents/Work/GenomeRunner/MDmisc/R/gene_enrichment_hg19.R")
entresIDs <- unique(as.data.frame(annotatedPeak[closest, ])$feature)
library(biomaRt)
mart <- useMart("ensembl", dataset="hsapiens_gene_ensembl")
genes <- getBM(attributes = "entrezgene", filters="ensembl_gene_id", values=entresIDs, mart = mart, uniqueRows = T)
```

MF

```{r eval=FALSE}
res <- Enrichment(as.character(unlist(genes)), id="entrezgene", use="GO", ont="MF")
write.xlsx(res, "/results/enrichment-DMR_CpG-global.xlsx", sheetName = "MF", row.names=F, append = T)
```

BP

```{r eval=FALSE}
res <- Enrichment(as.character(unlist(genes)), id="entrezgene", use="GO", ont="BP")
write.xlsx(res, "/results/enrichment-DMR_CpG-global.xlsx", sheetName = "BP", row.names=F, append = T)
```

KEGG

```{r eval=FALSE}
res <- Enrichment(as.character(unlist(genes)), id="entrezgene", use="KEGG")
write.xlsx(res, "/results/enrichment-DMR_CpG-global.xlsx", sheetName = "KEGG", row.names=F, append = T)
```

Reactome

```{r eval=FALSE}
res <- enrichPathway(gene = as.character(unlist(genes)), organism = "human", pAdjustMethod = "BH", pvalueCutoff = 0.1, qvalueCutoff = 0.2, readable = TRUE)
write.xlsx(summary(res), "/results/enrichment-DMR_CpG-global.xlsx", sheetName = "Reactome",row.names=FALSE,  append = T)
```

DO

```{r eval=FALSE}
res <- enrichDO(gene = as.character(unlist(genes)), ont = "DO", pAdjustMethod = "none", pvalueCutoff = 0.1, qvalueCutoff = 0.2, readable = TRUE)
write.xlsx(summary(res), "/results/enrichment-DMR_CpG-global.xlsx", sheetName = "DO",row.names=FALSE,  append = T)
```

09/10/2015 - Global DMRs separated into hyper- and hypomehtylated
===

`dmr_all.Rdata` - Attached is the significant (FDR < 1%) globally changed DMR findings with the directional annotations:

- num_cpg is the number of CpGs in the DMR
- pos_coef is the number of CpGs with a positive coefficient (hyper in visit 2)
- neg_coef is the number of CpGs with a negative coefficient (hypo in visit 2)

The breakdown of the 10,498 significant DMRs is summarized in the table below. I think interesting that for the 3rd category, where the DMR contains a mix of hyper and hypo CpGs, the majority of these DMRs have more hypo CpGs (2860 vs 1400).

- DMR.type	N
- CpG all Hyper	3067
- CpG all Hypo	3171
- CpG mixed	4260

```{r}
load("data/Tim/dmr_all.Rdata")
# Check that the counts match
nrow(temp[temp$pos_coef > 0 & temp$neg_coef == 0,]) # CpG all Hyper	3067
nrow(temp[temp$pos_coef == 0 & temp$neg_coef > 0,]) # CpG all Hypo	3171
nrow(temp[temp$pos_coef > 0 & temp$neg_coef > 0,]) # CpG mixed	4260
```

Save the data into BED files

```{r}
temp.all <- temp[temp$pos_coef > 0 & temp$neg_coef == 0,] # CpG all Hyper	3067
write.table(data.frame(paste("chr", temp.all$chr, sep=""), temp.all$start, temp.all$stop), "/Users/mikhail/Documents/Work/VCU_work/Coleen/Breast_cancer/data/Tim/DMR_CpG_hyper.bed", sep="\t", row.names=FALSE, col.names=FALSE, quote=F)

temp.all <- temp[temp$pos_coef == 0 & temp$neg_coef > 0,] # CpG all Hypo	3171
write.table(data.frame(paste("chr", temp.all$chr, sep=""), temp.all$start, temp.all$stop), "/Users/mikhail/Documents/Work/VCU_work/Coleen/Breast_cancer/data/Tim/DMR_CpG_hypo.bed", sep="\t", row.names=FALSE, col.names=FALSE, quote=F)

temp.all <- temp[temp$pos_coef > 0 & temp$neg_coef > 0,] # CpG mixed	4260
write.table(data.frame(paste("chr", temp.all$chr, sep=""), temp.all$start, temp.all$stop), "/Users/mikhail/Documents/Work/VCU_work/Coleen/Breast_cancer/data/Tim/DMR_CpG_hyperhypo.bed", sep="\t", row.names=FALSE, col.names=FALSE, quote=F)
```

Start the enrichment analysis 

```{r}
library(ChIPpeakAnno)
data(TSS.human.NCBI36)
library(org.Hs.eg.db)
```

DMRs-global-hyper

```{r}
group1.bed <- read.table("data/Tim/DMR_CpG_hyper.bed", sep="\t", stringsAsFactors = FALSE)
group1.bed <- BED2RangedData(group1.bed) # Convert to RangedData
annotatedPeak = annotatePeakInBatch(group1.bed, AnnotationData=TSS.human.NCBI36, select = "first")
over <- getEnrichedGO(annotatedPeak, orgAnn="org.Hs.eg.db", maxP=0.01, multiAdj=FALSE, minGOterm=10, multiAdjMethod="")
write.xlsx(unique(over$bp[, c(-10)]), "/results/enrichment-DMR_CpG-global.xlsx", sheetName = "BP-peaks-hyper", row.names=F, append = T)
write.xlsx(unique(over$mf[, c(-10)]), "/results/enrichment-DMR_CpG-global.xlsx", sheetName = "MF-peaks-hyper", row.names=F, append = T)
write.xlsx(unique(over$cc[, c(-10)]), "/results/enrichment-DMR_CpG-global.xlsx", sheetName = "CC-peaks-hyper", row.names=F, append = T)
enrichedPATH <- getEnrichedPATH(annotatedPeak, orgAnn= 'org.Hs.eg.db', pathAnn='KEGG.db', feature_id_type="ensembl_gene_id", maxP=0.01, minPATHterm=10, multiAdjMethod="none")
write.xlsx(unique(enrichedPATH[, c(-2)]), "/results/enrichment-DMR_CpG-global.xlsx", sheetName = "KEGG-peaks-hyper", row.names=F, append = T)
```

DMRs-global-hypo

```{r}
group1.bed <- read.table("data/Tim/DMR_CpG_hypo.bed", sep="\t", stringsAsFactors = FALSE)
group1.bed <- BED2RangedData(group1.bed) # Convert to RangedData
annotatedPeak = annotatePeakInBatch(group1.bed, AnnotationData=TSS.human.NCBI36, select = "first")
over <- getEnrichedGO(annotatedPeak, orgAnn="org.Hs.eg.db", maxP=0.01, multiAdj=FALSE, minGOterm=10, multiAdjMethod="")
write.xlsx(unique(over$bp[, c(-10)]), "/results/enrichment-DMR_CpG-global.xlsx", sheetName = "BP-peaks-hypo", row.names=F, append = T)
write.xlsx(unique(over$mf[, c(-10)]), "/results/enrichment-DMR_CpG-global.xlsx", sheetName = "MF-peaks-hypo", row.names=F, append = T)
write.xlsx(unique(over$cc[, c(-10)]), "/results/enrichment-DMR_CpG-global.xlsx", sheetName = "CC-peaks-hypo", row.names=F, append = T)
enrichedPATH <- getEnrichedPATH(annotatedPeak, orgAnn= 'org.Hs.eg.db', pathAnn='KEGG.db', feature_id_type="ensembl_gene_id", maxP=0.01, minPATHterm=10, multiAdjMethod="none")
write.xlsx(unique(enrichedPATH[, c(-2)]), "/results/enrichment-DMR_CpG-global.xlsx", sheetName = "KEGG-peaks-hypo", row.names=F, append = T)
```

DMRs-global-hyperhypo

```{r}
group1.bed <- read.table("data/Tim/DMR_CpG_hyperhypo.bed", sep="\t", stringsAsFactors = FALSE)
group1.bed <- BED2RangedData(group1.bed) # Convert to RangedData
annotatedPeak = annotatePeakInBatch(group1.bed, AnnotationData=TSS.human.NCBI36, select = "first")
over <- getEnrichedGO(annotatedPeak, orgAnn="org.Hs.eg.db", maxP=0.01, multiAdj=FALSE, minGOterm=10, multiAdjMethod="")
write.xlsx(unique(over$bp[, c(-10)]), "/results/enrichment-DMR_CpG-global.xlsx", sheetName = "BP-peaks-hyperhypo", row.names=F, append = T)
write.xlsx(unique(over$mf[, c(-10)]), "/results/enrichment-DMR_CpG-global.xlsx", sheetName = "MF-peaks-hyperhypo", row.names=F, append = T)
write.xlsx(unique(over$cc[, c(-10)]), "/results/enrichment-DMR_CpG-global.xlsx", sheetName = "CC-peaks-hyperhypo", row.names=F, append = T)
enrichedPATH <- getEnrichedPATH(annotatedPeak, orgAnn= 'org.Hs.eg.db', pathAnn='KEGG.db', feature_id_type="ensembl_gene_id", maxP=0.01, minPATHterm=10, multiAdjMethod="none")
write.xlsx(unique(enrichedPATH[, c(-2)]), "/results/enrichment-DMR_CpG-global.xlsx", sheetName = "KEGG-peaks-hyperhypo", row.names=F, append = T)
```
